lib/std/log.zig

```diff
@@ -0,0 +1,74 @@
+const std = @import("std.zig");
+const debug = std.debug;
+const builtin = std.builtin;
+const root = @import("root");
+
+pub const Level = enum {
+    /// helpful information for debugging.
+    Info,
+    /// can't be sure something is wrong, but it is probably worth investigating
+    Warning,
+    /// a bug in the application has been detected, but it is recoverable.
+    Error,
+    /// a bug in the application has been detected, and the application must
+    /// terminate now.
+    Fatal,
+};
+
+/// The default log level is based on build mode. Note that in ReleaseSmall
+/// builds logging is totally disabled (log.fatal still causes a panic though).
+pub const default_level: Level = switch (builtin.mode) {
+    .Debug => .Info,
+    .ReleaseSafe => .Warning,
+    .ReleaseFast => .Error,
+    .ReleaseSmall => .Fatal,
+};
+
+pub const level: Level = if (@hasDecl(root, "log_level")) root.log_level else default_level;
+
+fn log(message_level: Level, comptime format: []const u8, args: var) void {
+    if (@enumToInt(message_level) < @enumToInt(level))
+        return;
+    if (@hasDecl(root, "log")) {
+        root.log(message_level, format, args);
+    } else {
+        const held = debug.getStderrMutex().acquire();
+        defer held.release();
+        const stderr = debug.getStderrStream();
+        nosuspend stderr.print(fmt, args) catch return;
+    }
+}
+
+/// Log a message to stderr if log.level is Info. This log level is intended to
+/// be used for any information about the state of the program that would be
+/// useful while debugging.
+pub fn info(comptime format: []const u8, args: var) void {
+    log(.Info, format, args);
+}
+
+/// Log a message to stderr if log.level is Info or Warning. This log level is
+/// intended be used when it is uncertain if something has gone wrong or not,
+/// but the circumstances are worth investigating.
+pub fn warn(comptime format: []const u8, args: var) void {
+    log(.Warning, format, args);
+}
+
+/// Log a message to stderr if log.level is Info, Warning, or Error. This log
+/// level is intended be used when an error has occurred but the program is
+/// able to recover and continue running.
+pub fn err(comptime format: []const u8, args: var) void {
+    @setCold(true);
+    log(.Error, format, args);
+}
+
+/// Log a message to stderr if log.level is Info, Warning, Error, or Fatal.
+/// This log level should only be used for irrecoverable errors as it will
+/// trigger a panic after logging. Note that this will not log if the build
+/// mode is release small but will still trigger a panic.
+pub fn fatal(comptime format: []const u8, args: var) noreturn {
+    @setCold(true);
+    // On release small builds we don't log because that would require storing
+    // the error message.
+    if (!builtin.mode == .ReleaseSmall) log(.Fatal, format, args);
```

Comment:
looks like a typo, should be `!=` ?

Reply:
indeed, i seem to have placed the `!` before `builtin.mode` for some reason.